kcidb/io/schema/v3.py

```diff
@@ -0,0 +1,626 @@
+# commit hash (sha1 or sha256)
+GIT_REPOSITORY_COMMIT_HASH_PATTERN = f"({SHA1_PATTERN}|{SHA256_PATTERN})"
+
+# A regular expression pattern matching strings containing revision IDs
+REVISION_ID_PATTERN = \
+    f"{GIT_REPOSITORY_URL_PATTERN}" \
+    f"@{GIT_REPOSITORY_COMMIT_HASH_PATTERN}" \
+    f"(\\+{SHA256_PATTERN})?"
+
+# A regular expression pattern matching strings containing origin name
+ORIGIN_PATTERN = "[a-z0-9_]+"
+
+# A regular expression pattern matching strings containing an object ID
+ORIGIN_ID_PATTERN = f"{ORIGIN_PATTERN}:.*"
+
+# JSON schema for a named remote resource
+JSON_RESOURCE = {
+    "title": "resource",
+    "description": "A named remote resource",
+    "type": "object",
+    "properties": {
+        "name": {
+            "type": "string",
+            "description":
+                "Resource name. Must be usable as a local file name for the "
+                "downloaded resource. Cannot be empty. Should not include "
+                "directories.",
+            "pattern": "^[^/]+$",
+        },
+        "url": {
+            "type": "string",
+            "format": "uri",
+            "description":
+                "Resource URL. Must point to the resource file directly, "
+                "so it could be downloaded automatically.",
+        },
+    },
+    "additionalProperties": False,
+    "required": [
+        "name",
+        "url",
+    ],
+    "examples": [
+        {
+            "name": "console.log",
+            "url":
+                "https://artifacts.cki-project.org/pipelines/223563/logs/"
+                "aarch64_host_1_console.log"
+        },
+        {
+            "name": "kernel.tar.gz",
+            "url":
+                "https://artifacts.cki-project.org/pipelines/224569/"
+                "kernel-stable-aarch64-"
+                "a2fc8ee6676067f27d2f5c6e4d512adff3d9938c.tar.gz"
+        }
+    ]
+}
+
+# JSON schema for a code revision
+JSON_REVISION = {
+    "title": "revision",
+    "description":
+        "A revision of the tested code.\n"
+        "\n"
+        "Represents a way the tested source code could be obtained. E.g. "
+        "checking out a particular commit from a git repo, and applying a "
+        "set of patches on top.",
+    "type": "object",
+    "properties": {
+        "id": {
+            "type": "string",
+            "description":
+                "Revision ID.\n"
+                "\n"
+                "Must be a string containing the following, in order.\n"
+                "\n"
+                "The URL of the Git repository which contains the base code "
+                "of the revision. The shortest possible https:// URL, or, if "
+                "that's not available, the shortest possible git:// URL."
+                "\n"
+                "The '@' character.\n"
+                "\n"
+                "The full commit hash of the revision's base code in the Git "
+                "repository.\n"
```

Comment:
It would seem wise to specify that the commit has should be SHA-1 since there are talks of using SHA-256.  Otherwise, if different sources use different SHA versions then we won't be able to join the data (unless we have a way to determine SHA-1 / SHA-256 commit equivalents, which would seem over-the-top).

Reply:
I think if the kernel development switches to SHA-256, then all the forks will have to switch, wouldn't they? And then we won't have to join any data.

Reply:
I'm not entirely sure but I think git would allow doing a pull from a repository which doesn't have SHA-256 hashes into one that does.  I guess the real question is whether new versions of git will keep both SHA-1 and SHA-256 hashes for a same commit, in which case we could have a problem.  I believe it won't need to do that in which case we're probably fine with allowing both.  It would still be good to actually prove that before allowing SHA-256 imho.

Reply:
Alright, allowing SHA-256 might be a little premature and we can always permit it later, without even bumping the major version number of the schema. Removed.